File: itx-rs/itx-impl/src/queue/sqs/factory.rs
use std::sync::Arc;

use aws_sdk_sqs::Client;
use itx_contract::queue::MessageQueue;
use itx_contract::queue::factory::MessageQueueFactory;

use crate::queue::sqs::SqsMessageQueue;

#[derive(Clone, serde::Deserialize)]
#[serde(rename_all = "kebab-case")]
pub struct SqsMessageQueueFactoryProps {
    pub local_endpoint_url: Option<String>,
    pub max_concurrency: u32,
    pub control_standard_queue_url: String,
    pub control_premium_queue_url: String,
    pub compute_standard_queue_url: String,
    pub compute_premium_queue_url: String,
}

pub struct SqsMessageQueueFactory {
    client: Client,
    props: SqsMessageQueueFactoryProps,
}

impl SqsMessageQueueFactory {
    pub async fn new(props: SqsMessageQueueFactoryProps) -> Self {
        let aws_config = aws_config::load_from_env().await;
        let mut sqs_config = aws_sdk_sqs::config::Builder::from(&aws_config);
        if let Some(endpoint) = &props.local_endpoint_url {
            sqs_config = sqs_config.endpoint_url(endpoint);
        }
        let client = Client::from_conf(sqs_config.build());

        Self { client, props }
    }
}

impl MessageQueueFactory for SqsMessageQueueFactory {
    fn create_control_standard_queue(&self) -> Arc<dyn MessageQueue> {
        Arc::new(SqsMessageQueue::new(
            self.client.clone(),
            self.props.control_standard_queue_url.clone(),
            self.props.max_concurrency,
        ))
    }

    fn create_control_premium_queue(&self) -> Arc<dyn MessageQueue> {
        Arc::new(SqsMessageQueue::new(
            self.client.clone(),
            self.props.control_premium_queue_url.clone(),
            self.props.max_concurrency,
        ))
    }

    fn create_compute_standard_queue(&self) -> Arc<dyn MessageQueue> {
        Arc::new(SqsMessageQueue::new(
            self.client.clone(),
            self.props.compute_standard_queue_url.clone(),
            self.props.max_concurrency,
        ))
    }

    fn create_compute_premium_queue(&self) -> Arc<dyn MessageQueue> {
        Arc::new(SqsMessageQueue::new(
            self.client.clone(),
            self.props.compute_premium_queue_url.clone(),
            self.props.max_concurrency,
        ))
    }
}

File: itx-go/itx-impl/queue/sqs/factory.go
package sqs

import (
	"context"

	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
	"github.com/chehsunliu/itx/itx-go/itx-contract/queue"
)

type FactoryProps struct {
	LocalEndpointURL        string `yaml:"local-endpoint-url"`
	MaxConcurrency          int64  `yaml:"max-concurrency"`
	ControlStandardQueueURL string `yaml:"control-standard-queue-url"`
	ControlPremiumQueueURL  string `yaml:"control-premium-queue-url"`
	ComputeStandardQueueURL string `yaml:"compute-standard-queue-url"`
	ComputePremiumQueueURL  string `yaml:"compute-premium-queue-url"`
}

type MessageQueueFactory struct {
	client *sqs.Client
	props  FactoryProps
}

func New(props FactoryProps) (*MessageQueueFactory, error) {
	cfg, err := config.LoadDefaultConfig(context.Background())
	if err != nil {
		return nil, err
	}

	client := sqs.NewFromConfig(cfg, func(o *sqs.Options) {
		if props.LocalEndpointURL != "" {
			o.BaseEndpoint = &props.LocalEndpointURL
		}
	})
	return &MessageQueueFactory{client: client, props: props}, nil
}

func (f *MessageQueueFactory) CreateControlStandardQueue() queue.MessageQueue {
	return newMessageQueue(f.client, f.props.ControlStandardQueueURL, f.props.MaxConcurrency)
}

func (f *MessageQueueFactory) CreateControlPremiumQueue() queue.MessageQueue {
	return newMessageQueue(f.client, f.props.ControlPremiumQueueURL, f.props.MaxConcurrency)
}

func (f *MessageQueueFactory) CreateComputeStandardQueue() queue.MessageQueue {
	return newMessageQueue(f.client, f.props.ComputeStandardQueueURL, f.props.MaxConcurrency)
}

func (f *MessageQueueFactory) CreateComputePremiumQueue() queue.MessageQueue {
	return newMessageQueue(f.client, f.props.ComputePremiumQueueURL, f.props.MaxConcurrency)
}
